029-across-step/proposal.md

```diff
@@ -0,0 +1,321 @@
+# `across` step
+
+This proposal introduces a mechanism by which a build plan can be executed
+multiple times with different `((vars))`, in service of build matxies, pipeline
+matrixes, and - building on [`((var))` sources][var-sources-rfc] - dynamic
+variations of both.
+
+
+## Motivation
+
+* Support dynamic multi-branch workflows:
+  [concourse/concourse#1172][multi-branch-issue]
+* Support build matrixes and pipeline matrixes.
+* Replace a common, and very flawed, usage pattern of `version: every`, which
+  has proven to be very complicated to support.
+
+
+## Proposal
+
+The `across` step modifier is given a list containing vars and associated
+values to set when executing the step. The step will be executed across all
+combinations of var values.
+
+### Static var values
+
+The set of values can be static, as below:
+
+```yaml
+task: unit
+vars: {go_version: ((.:go_version))}
+across:
+- var: go_version
+  values: [1.12, 1.13]
+```
+
+This will run the `unit` task twice: once with `go_version` set to `1.12`, and
+again with `1.13`.
+
+### Dynamic var values from a var source
+
+Rather than static values, var values may be pulled from a var source
+dynamically:
+
+```yaml
+var_sources:
+- name: booklit-prs
+  type: github-prs
+  config:
+    repository: vito/booklit
+    access_token: # ...
+
+plan:
+- set_pipeline: pr
+  instance_vars: {pr_number: ((.:pr.number))}
+  across:
+  - var: pr
+    source: booklit-prs
```

Comment:
Have you thought about allowing the `across` step to iterate over objects from an arbitrary prototype (i.e. not just constrained to `var_source` prototypes and their `list` message)? That could make the object streams returned by prototypes more useful outside of the special prototype interfaces that Concourse will interact with natively (resources, var_sources, notifications, etc.)

In this comment https://github.com/concourse/concourse/pull/5936#issuecomment-667192099, you suggested that the `run` step could possibly support a `set_var` field to store the returned object as a local variable. Perhaps an alternative to `set_var` would be to use the `across` step in its place, e.g.

```yaml
- across:
  - var: some-var
    type: gval
    params: {fields: {some-field: hello}}
  run: echo
  type: debug
  params: {message: ((.:some-var.some-field))}
```

Kind of weird in this case since we're iterating over a single value always - but I guess for `set_var` to work in general, it would also have to set the var to a list of objects (even if a single object is emitted)

Reply:
This also makes me realize that the `across` step can be used for conditional portions of build plans, by having the list of values be either 0 or 1 element long 🤔

Reply:
It's also making me wonder about dynamic build plans in the context of the `across` step (i.e. the list of values to iterate over can be determined at runtime) - what are your thoughts on that? I think it could greatly extend the use cases that `across` can support, but it could possibly make pipelines difficult to understand since you could introduce arbitrarily complicated control flow with the point above

Reply:
I have, and your comments are pretty much the same thought process I went thought. 😄 It seems like a really cool idea, but yeah, the difficulty is that it seems to require some sort of dynamic build plan construct.

I'm running into another area that would benefit from dynamic build plans: the way `file:` and `image_resource:` interact on a task step. I'm working on a refactor that pulls image fetching out into explicit `check` and `get` steps in the build plan, but the existence of `file:` means I can't know whether they're needed (Darwin and Windows have no images, for example), or what image to even `check` and `get`, until the build is running. In this case I can probably just preemptively configure steps which may just do no-ops, and they can probably get their config in a similar way to how the `Get` plan uses `VersionFrom` to get some config from a prior step, but it all feels kind of held together with duct tape. (There's also a bit of complexity in that we may need to `check` + `get` a custom type that the `image_resource:` uses, but I think have a way to get around that.)

I still don't see a super obvious way to implement dynamic build plans though. 🤔 But I haven't really tried, either, I guess.

Reply:
The only thought I had was for there to be some sort of build event that says "inject this build plan under this plan ID" or "replace this plan ID with this build plan". So, the plan would initially look something like:

```json
{
  "schema": "exec.v2",
  "plan": {
    "id": "across-id",
    "across": {}
  }
}
```

and there could be a build event like:

```json
{
  "event":"dynamic-build-plan",
  "version":"1.0",
  "data": {
    "id": "across-id",
    "across": {
      "some": "data"
    }
  }
}
```

which the UI would interpret to construct a final build plan of:

```json
{
  "schema": "exec.v2",
  "plan": {
    "id": "across-id",
    "across": {
      "some": "data"
    }
  }
}
```

I'm sure there are some technical challenges I haven't considered, though

Reply:
Ooh, I like the idea of emitting plans in the event stream - seems a lot simpler than e.g. modifying and re-saving the build plan directly.

It'd be a little interesting to figure out the best way to structure it, for example how to indicate how plan A is related to plan B so that the UI can know how these dynamic plans should be displayed. For my `task` step example from earlier, there may be N plans if we e.g. need to perform `check` and `get` of a full hierarchy of chained resource types leading up to the one used in `image_resource`. Maybe these could all be given some sort of 'image-fetching' tag or something. 🤔

Reply:
Just trying to understand the use case you're describing/the refactor you're working on - is it that whenever we use an image (be it in a `get`, `put`, `task`, or `check`) we would have a `check` + `get` for the (potential chain of) parent `resource_types` and for the anonymous `image_resource` for `task` steps - but the latter can only be determined at runtime?

Does it make sense for that "image fetching" plan to be a part of the plan for those step types? For instance, a `put` step could have a plan like this:

```json
{
  "id": "1",
  "put": {
    "name": "some-resource",
    "type": "some-resource-type",
    "image_fetching_plan": {
      "id": "2",
      "do": [
        {
          "id": "3",
          "check": {"type": "registry-image"}
        },
        {
          "id": "4",
          "get": {"type": "registry-image"}
        },
        {
          "id": "5",
          "check": {"type": "some-resource-type"}
        },
        {
          "id": "6",
          "get": {"type": "some-resource-type"}
        }
      ]
    }
  }
}
```

(I simplified what the plan would look like for the explicit `get` and `check` steps, since I'm not sure what they'll look like exactly. Also, does it apply to base resource types like `registry-image` as well?)

---

For `task` steps with a `file:`, we could initialize the `task` plan to:

```json
{
  "id": "1",
  "task": {
    "name": "some-task",
    "image_fetching_plan": {
      "id": "2",
      "do": []
    }
  }
}
```

and after we have the `file:` artifact, we can send a `dynamic-build-plan` event:

```json
{
  "event":"dynamic-build-plan",
  "version":"1.0",
  "data": {
    "id": "2",
    "do": [
        {
          "id": "3",
          "check": {"type": "registry-image"}
        },
        {
          "id": "4",
          "get": {"type": "registry-image"}
        },
        {
          "id": "5",
          "check": {"name": "some-image"}
        },
        {
          "id": "6",
          "get": {"name": "some-image"}
        }
    ]
  }
}
```

Does that make any sense?

Reply:
@aoldershaw Yeah that makes a lot of sense.

My current approach is as follows:

* Use `CheckPlan` to run a `check` for the image
* Use `VersionFrom` on the `GetPlan` to get the version returned from the `CheckPlan`
  * Normally `VersionFrom` this is used for the `get` after a `put`, but it seems to be a clean fit here too
* Add `ImageArtifactName` to all other core plans (`check`, `get`, `put`), and use to refer to the artifact fetched by the `GetPlan`
  * Normally `ImageArtifactName` is used by the `TaskPlan` for when [`image:`](https://concourse-ci.org/jobs.html#schema.step.task-step.image) is set on a `task` step.
* Remove all image resource `check`/`get` logic from `atc/worker/`, now that everything is either a base resource type or an already-fetched image artifact

So far I've just chained it all together using `OnSuccessPlan` like this:

```json
{
  "id": "1",
  "on_success": {
    "step": {
      "id": "check-type-id",
      "check": {
        "type": "registry-image",
        "source": {
          "repository":"concourse/git-resource"
        }
      }
    },
    "hook": {
      "id": "3",
      "on_success": {
        "step": {
          "id": "4",
          "get": {
            "name": "type:my-git",
            "type": "registry-image",
            "source": {
              "repository":"concourse/git-resource"
            },
            "version_from": "check-type-id"
          }
        },
        "hook": {
          "id": "5",
          "put": {
            "image_artifact_name": "type:my-git",
            "source": {
              "uri": "https://github.com/vito/booklit"
            }
          }
        }
      }
    }
  }
}
```

The downside of this approach is that I think the UI will literally render this as a bunch of `OnSuccess` hooks, which isn't ideal. I could use `do`, but that'll also look surprisingly different from what the user expects. So some approach like yours, which retains the original plan structure and instead adds the setup plans as sub-fields of the steps which need it, could be a better structure.

I haven't actually gotten this code to work yet - I immediately ran into a bit of a roadblock, as it's impossible to inter-mix `CheckPlan` with any other kind of plan. `CheckPlan`s can only be run as LIDAR checks, cannot be run within builds, and other types of plans cannot be added to LIDAR checking plans - it nil panics if the toplevel plan isn't a `CheckPlan`.

So now I'm going down a rabbit hole of having all `check`s run as builds instead of having their own separate stack. The idea is to end up with two build schedulers: a resource scheduler alongside the job scheduler. So far it looks promising; it feels kind of nice to have builds be the common substrate for everything that we run - it may be the same way that we approach Projects, which also execute build plans.

If you want to take a peek at what I have so far, the branch is [`checks-as-builds`](https://github.com/concourse/concourse/tree/checks-as-builds). So far it's super spiky - once I have more confidence in this being a good path to go down I plan to break it up into many smaller changes and be a bit more strategic about it.

Reply:
Follow-up: the dynamic sub-steps stuff was delivered in v7, which now runs `check`s as builds and also performs image fetching as sub- `check` and `get` steps! 🎉

We're using more of this dynamic sub-step stuff for https://github.com/concourse/concourse/issues/6274

Given this new capability it's actually pretty tempting to re-frame this as something like this:

```yml
- get_vars: foos
  trigger: true
- across:
  - var: foo
    values: ((.:foos))
  do: # ...
```

It's a bit more verbose, but I like that it lets you trigger off of changes to the set without having to `across:` something. This could be used for re-building a website based on the set of PRs, for example, instead of having to iterate over them in the build plan. (I have this use case right now while I'm working on adding an RFC status board to the website. 😄)

...actually, what if we just made the `across:` value be a list of steps to run, and then we go over all of their results? We could make `values:` be an actual step that runs and returns the predefined list of values, and then you could do `run:` or `get_vars:` or whatever you want too.

```yml
across:
- var: foo
  get_vars: foos
  trigger: true
- var: foo
  values: [1, 2, 3]
- var: foo
  run: blah
```

Reply:
I like that idea! Small suggestion - what if instead of inlining in the var definition, we supported `values` and `values_from`, e.g.:

```yaml
across:
- var: foo
  values_from:
    get_vars: foos
    trigger: true
- var: bar
  values: [1, 2, 3]
```

I feel like that's maybe a little more obvious/readable at a glance.

Also, I feel like the name `get_vars` is a little awkward, especially because (and correct me if I'm wrong) the `list` operation of `var_sources` doesn't necessarily give the "values" of the vars (it'd be a bit risky for a `vault` var_source to potentially leak all credentials, even if you have access to them). Maybe it could be something like `list_vars`?

---

Here's a bit of a wild idea that poses several complications (and I don't even think is a good idea really, kinda just spitballing) - but what if `values_from` could be any step that produced a stream of values. This could be `get_vars`, `load_var`, `run`, or a `do` where the last step produces a stream. Then, you could do stuff like:

```yaml
- get: ci
- across:
  - var: foo
    values_from:
      do:
      - get: data
        trigger: true
      - task: filter-list
        file: ci/tasks/filter-list.yml
        input_mapping: {input: data}
        output_mapping: {output: filtered}
      - load_var: filtered
        file: filtered/data.json
```

Which would then imply 2+ levels of checking/triggering - one level triggers on changes to the `trigger` resources/vars (which runs a `check` or `get_var`), and on changes to the resource/var, we trigger a build that runs the full `do` plan + any `get` steps in advance (perhaps we require inputs to be explicitly specified in the case of a `task`). e.g. in this case, the plan would be something like:

```yaml
- get: ci
- get: data
- task: filter-list
  ...
- load_var: filtered
  ...
```

...and when `filtered` changes, we trigger the rest of the parent build plan

Again, don't think this is really a good idea haha, just throwing it out there

Reply:
That seems close to what I'm proposing - whatever the step results in gets used as the value. I think it'd just look like this with the syntax I had before:

```yaml
var: foo
do:
- get: data
  trigger: true
- task: filter-list
  file: ci/tasks/filter-list.yml
  input_mapping: {input: data}
  output_mapping: {output: filtered}
- load_var: filtered
  file: filtered/data.json
```

...but this is just a different syntax. `values_from` is probably clearer. I think we'd still need to handle `values: ((.:foo))`, but that could probably be handled by the `across` step too instead of a separate `values:` step.

In either case, I'm not sure how the `across` step would access the result in the case of a `do`. 🤔 Currently results are stored under step IDs stored in the `RunState`, but this is almost like a return value from an expression. With a single step it's obvious what step ID to use, but with a `do` it's not so easy. It's almost like we need `(exec.Step).Run` to be able to return a value on its own, and `do` would return the last value. But that changes the interface.

Maybe this whole storing-results-in-`RunState` pattern could/should change? @clarafu @chenbh Has your recent experience in this area led to a similar line of thinking?

Reply:
Oh, also re: `get_vars` vs `list_vars`, yeah depending on what the resulting value is `list_vars` might be more accurate. I think either `get_vars` would actually fetch all the var values individually, using a `list` call under the hood for the scheduling/triggering, or we would have `list_vars` instead and allow the user to run a separate `get_var` step for each var value if that's desired. Don't know which; haven't put much thought into it yet.

Reply:
What I was proposing is a bit more complicated than just a multi-step `values_from` (where the value comes from the last step), but I didn't have the proper framing. Now, I realize it the proposal was actually a combination of two ideas:

1. A new way to communicate between jobs by emitting/receiving data
2. Anonymous jobs (run nested in other jobs)

I'll flesh out my original example a bit more:

```yaml
jobs:
- name: parent
  plan:
  - get: ci
  - across:
    - var: foo
      values_from:
        do:  # <----------------------------- A
        - get: data
          trigger: true <-------------------- B
        - task: filter-list
          file: ci/tasks/filter-list.yml
          input_mapping: {input: data}
          output_mapping: {output: filtered}
        - load_var: filtered # <------------- C
          file: filtered/data.json
    ...
```

Here, A is an "anonymous job". When the trigger on B gets a new version, the `parent` job doesn't trigger - only the anonymous job A. The `load_var` C implicitly emits the value it loads, and `parent` implicitly receives and triggers on changes to this value.

The implicit emitting and receiving is very non-obvious, though, and the anonymous job idea is a bit odd (either we make it fully self-contained and need to duplicate the `get: ci`, or it magically can borrow artifacts from the parent job). What if instead we made it explicit and gave the filtering it's own job:

```yaml
jobs:
- name: filter-list
  plan:
  - get: ci
  - get: data
    trigger: true
  - task: filter-list
    file: ci/tasks/filter-list.yml
    input_mapping: {input: data}
    output_mapping: {output: filtered}
  - load_var: filtered
    file: filtered/data.json
    emit: data # <--------------------- A

- name: parent
  plan:
  - get: ci
  - receive: data <-------------------- B
    job: filter-list
    trigger: true
  - across:
    - var: foo
      values: ((.:data))
    ...
```

A emits a new value and stores it as `data`. B triggers on changes to that emitted value and stores the result in a local variable.

This mechanism could even be used for `{get_var: ..., trigger: true}` - the builds we run periodically to fetch var could `emit` the value to the DB, and the job would `receive` that changed value from the build.

Reply:
> In either case, I'm not sure how the across step would access the result in the case of a do. 🤔 Currently results are stored under step IDs stored in the RunState, but this is almost like a return value from an expression. With a single step it's obvious what step ID to use, but with a do it's not so easy. It's almost like we need (exec.Step).Run to be able to return a value on its own, and do would return the last value. But that changes the interface.

I thought of it like every type of step would have a single output (e.g. `check` is version, `get` is image spec, `load_var` and `get_var` are var), in `do`'s case maybe we just need to figure out the contract for what it should return (is it time to introduce a `result` step/modifier)? 

It would then be up to `across` to figure out what to do with this info (which might be wild cause that implies we can run `across` a list of versions from a `check`).

Reply:
I am not sure if I am adding anything useful to the conversation, but when I saw the the 7.0.0 release notes I immediately saw a good use case for one of our pipelines and I tried to build a working prototype using pipeline instances and the across step.

```yaml
- name: set-pipeline-instances
  serial: true
  plan:
    - in_parallel:
      - get: ci
      - get: src-branches
        trigger: true
    ...
    - load_var: branches
      file: branches-json/branches.json
    - set_pipeline: {{ (datasource "config").name }}
      file: ci/ci/build.yml
      instance_vars: {branch: ((.:b))}
      across:
        - var: b
          values: ((.:branches))
```

I am keen for the normal resource type to trigger the job and then use the `load_var` to load a variable that is a list and then just use that for for the argument to `values`.